File: src/main/java/com/cse3063f25grp1/orchestrator/RagPipeline.java
package com.cse3063f25grp1.orchestrator;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.cse3063f25grp1.answer.AnswerAgent;
import com.cse3063f25grp1.answer.TemplateAnswerAgent;
import com.cse3063f25grp1.config.Config;
import com.cse3063f25grp1.context.Context;
import com.cse3063f25grp1.data.ChunkLoader;
import com.cse3063f25grp1.data.ChunkStore;
import com.cse3063f25grp1.intent.IntentDetector;
import com.cse3063f25grp1.intent.IntentRulesLoader;
import com.cse3063f25grp1.intent.RuleIntentDetector;
import com.cse3063f25grp1.model.Answer;
import com.cse3063f25grp1.model.Hit;
import com.cse3063f25grp1.model.Intent;
import com.cse3063f25grp1.reranker.Reranker;
import com.cse3063f25grp1.reranker.SimpleReranker;
import com.cse3063f25grp1.retrieval.KeywordRetriever;
import com.cse3063f25grp1.retrieval.Retriever;
import com.cse3063f25grp1.trace.TraceBus;
import com.cse3063f25grp1.trace.TraceEvent;
import com.cse3063f25grp1.writer.HeuristicQueryWriter;
import com.cse3063f25grp1.writer.QueryWriter;

public abstract class RagPipeline {

    protected final Config config;
    protected final Context context;
    protected final com.cse3063f25grp1.trace.TraceBus traceBus;
    protected IntentDetector intentDetector;
    protected QueryWriter queryWriter;
    protected Retriever retriever;
    protected Reranker reranker;
    protected AnswerAgent answerAgent;

    public RagPipeline(Config config, Context context, TraceBus traceBus) {
        this.config = config;
        this.context = context;
        this.traceBus = traceBus;
    }

    public abstract void execute();

    protected void detectIntent() {
        long startTime = System.currentTimeMillis();
        String question = context.getQuestion().getText();
        String inputs = "question=\"" + question + "\"";
        String outputsSummary = "";
        String error = null;

        try {
            if (config.getIntentType().equals("RuleIntentDetector")) {
                IntentRulesLoader rulesLoader = new IntentRulesLoader();
                Map<Intent, List<String>> rules = rulesLoader.loadRules(config.getRulesFilePath());
                List<Intent> priorityOrder = new java.util.ArrayList<>(rules.keySet());
                intentDetector = new RuleIntentDetector(rules, priorityOrder);
            } else {
                throw new IllegalArgumentException("Unknown intent detector type: " + config.getIntentType());
            }
            Intent intent = intentDetector.detect(question);
            context.setIntent(intent);
            outputsSummary = "intent=" + intent;
        } catch (Exception e) {
            error = e.getMessage();
            throw e;
        } finally {
            long timingMs = System.currentTimeMillis() - startTime;
            traceBus.publish(new TraceEvent("detectIntent", inputs, outputsSummary, timingMs, error));
        }
    }

    protected void writeQuery() {
        long startTime = System.currentTimeMillis();
        String question = context.getQuestion().getText();
        String inputs = "";
        String outputsSummary = "";
        String error = null;

        try {
            if (config.getWriterType().equals("HeuristicQueryWriter")) {
                Set<String> stopwords = loadStopwords(config.getStopwordsFilePath());
                inputs = "stopwords=" + stopwords.size() + " stopwords";
                queryWriter = new HeuristicQueryWriter(stopwords);
            } else {
                throw new IllegalArgumentException("Unknown query writer type: " + config.getWriterType());
            }

            List<String> terms = queryWriter.write(question);
            context.setTerms(terms);
            outputsSummary = "Number of terms: "+ terms.size()+" Terms:"+terms.toString();
        } catch (Exception e) {
            error = e.getMessage();
            throw e;
        } finally {
            long timingMs = System.currentTimeMillis() - startTime;
            traceBus.publish(new TraceEvent("writeQuery", inputs, outputsSummary, timingMs, error));
        }
    }

    protected void retrieve() {
        long startTime = System.currentTimeMillis();
        List<String> terms = context.getTerms();
        String inputs = terms.toString();
        String outputsSummary = "";
        String error = null;

        try {
            if (config.getRetrieverType().equals("KeywordRetriever")) {
                retriever = new KeywordRetriever(config.getTopK());
            } else {
                throw new IllegalArgumentException("Unknown retriever type: " + config.getRetrieverType());
            }

            List<Hit> hits = retriever.retrieve(terms, context.getChunkStore());
            context.setRetrievedHits(hits);
            outputsSummary = "Number of hits: "+ hits.size()+" retrievedHits: "+hits.toString();
        } catch (Exception e) {
            error = e.getMessage();
            throw e;
        } finally {
            long timingMs = System.currentTimeMillis() - startTime;
            traceBus.publish(new TraceEvent("retrieve", inputs, outputsSummary, timingMs, error));
        }
    }

    protected void rerank() {
        long startTime = System.currentTimeMillis();
        List<String> terms = context.getTerms();
        List<Hit> hits = context.getRetrievedHits();
        String inputs = hits.toString();
        String outputsSummary = "";
        String error = null;

        try {
            if (config.getReranker().equals("SimpleReranker")) {
                int proximityWindow = 15;
                int proximityBonus = 5;
                int titleBoost = 3;
                reranker = new SimpleReranker(proximityWindow, proximityBonus, titleBoost);
            } else {
                throw new IllegalArgumentException("Unknown reranker type: " + config.getReranker());
            }

            ChunkLoader chunkLoader = new ChunkLoader();
            ChunkStore chunkStore = chunkLoader.loadChunks(config.getChunkPath());

            List<Hit> rerankedHits = reranker.rerank(terms, hits, chunkStore);
            context.setRerankedHits(rerankedHits);
            outputsSummary = rerankedHits.toString();
        } catch (Exception e) {
            error = e.getMessage();
            throw e;
        } finally {
            long timingMs = System.currentTimeMillis() - startTime;
            traceBus.publish(new TraceEvent("rerank", inputs, outputsSummary, timingMs, error));
        }
    }

    protected void answer() {
        long startTime = System.currentTimeMillis();
        List<Hit> rerankedHits = context.getRerankedHits();
        String inputs ="Number of hits: "+ rerankedHits.size() +" rerankedHits:"+rerankedHits.toString();
        String outputsSummary = "";
        String error = null;

        try {
            if (config.getAnswerAgentType().equals("TemplateAnswerAgent")) {
                answerAgent = new TemplateAnswerAgent();
            } else {
                throw new IllegalArgumentException("Unknown answer agent type: " + config.getAnswerAgentType());
            }

            // Use the configured AnswerAgent to generate a real answer
            List<String> queryTerms = context.getTerms();
            Answer answer = answerAgent.answer(queryTerms, rerankedHits, context.getChunkStore());
            context.setFinalAnswer(answer);
            outputsSummary = "Answer: "+answer.toString();
        } catch (Exception e) {
            error = e.getMessage();
            throw e;
        } finally {
            long timingMs = System.currentTimeMillis() - startTime;
            traceBus.publish(new TraceEvent("answer", inputs, outputsSummary, timingMs, error));
        }
    }

    // Helper method: Stopwords yükle
    protected Set<String> loadStopwords(Path stopwordsPath) {
        try {
            List<String> lines = Files.readAllLines(stopwordsPath);
            Set<String> stopwords = new HashSet<>();
            boolean inStopWords = false;

            for (String line : lines) {
                String trimmed = line.trim();
                if (trimmed.isEmpty() || trimmed.startsWith("#")) {
                    continue;
                }
                if (trimmed.equals("stop_words:")) {
                    inStopWords = true;
                    continue;
                }
                if (inStopWords && trimmed.startsWith("-")) {
                    String word = trimmed.substring(1).trim();
                    if (word.startsWith("\"") && word.endsWith("\"")) {
                        word = word.substring(1, word.length() - 1);
                    }
                    stopwords.add(word);
                }
            }
            return stopwords;
        } catch (Exception e) {
            throw new RuntimeException("Failed to load stopwords from: " + stopwordsPath, e);
        }
    }
}

File: iteration1_py/src/orchestrator/rag_pipeline.py
from abc import ABC, abstractmethod
from typing import List, Set, Dict
from pathlib import Path
import time

from src.config.config import Config
from src.context.context import Context
from src.trace.trace_bus import TraceBus
from src.trace.trace_event import TraceEvent
from src.model.intent import Intent
from src.model.hit import Hit
from src.model.answer import Answer


class RagPipeline(ABC):
    
    def __init__(self, config: Config, context: Context, trace_bus: TraceBus):
        
        self.__config = config
        self.__context = context
        self.__trace_bus = trace_bus
        
        self.__intent_detector = None
        self.__query_writer = None
        self.__retriever = None
        self.__reranker = None
        self.__answer_agent = None
    
    @abstractmethod
    def execute(self) -> None:
        pass
    
    def detect_intent(self) -> None:
        start_time = time.time()
        question = self.__context.get_question().get_text()
        inputs = f'question="{question}"'
        outputs_summary = ""
        error = None
        
        try:
            from src.intent.intent_rules_loader import IntentRulesLoader
            from src.intent.rule_intent_detector import RuleIntentDetector
            
            if self.__config.get_intent_type() == "RuleIntentDetector":
                rules_loader = IntentRulesLoader()
                rules = rules_loader.load_rules(self.__config.get_rules_file_path())
                priority_order = list(rules.keys())
                self._intent_detector = RuleIntentDetector(rules, priority_order)
                self.__context.set_intent_keyword_rules(rules)
            else:
                raise IllegalArgumentError(f"Unknown intent detector type: {self.__config.get_intent_type()}")
            
            intent = self._intent_detector.detect(question)
            self.__context.set_intent(intent)
            outputs_summary = f"intent={intent}"
        except Exception as e:
            error = str(e)
            raise
        finally:
            timing_ms = (time.time() - start_time) * 1000
            self.__trace_bus.publish(TraceEvent("detectIntent", inputs, outputs_summary, timing_ms, error))
    
    def write_query(self) -> None:
        start_time = time.time()
        question = self.__context.get_question().get_text()
        inputs = ""
        outputs_summary = ""
        error = None
        
        try:
            from src.writer.heuristic_query_writer import HeuristicQueryWriter
            
            if self.__config.get_writer_type() == "HeuristicQueryWriter":
                stopwords = self.load_stopwords(self.__config.get_stopwords_file_path())
                inputs = f"stopwords={len(stopwords)} stopwords{stopwords}"
                boosters = self.__context.get_intent_keyword_rules()
                if boosters is None:
                    boosters = {}
                self._query_writer = HeuristicQueryWriter(stopwords, boosters)
            else:
                raise IllegalArgumentError(f"Unknown query writer type: {self.__config.get_writer_type()}")
            
            terms = self._query_writer.write(question, self.__context.get_intent())
            self.__context.set_terms(terms)
            outputs_summary = f"Number of terms: {len(terms)} Terms:{terms}"
        except Exception as e:
            error = str(e)
            raise
        finally:
            timing_ms = (time.time() - start_time) * 1000
            self.__trace_bus.publish(TraceEvent("writeQuery", inputs, outputs_summary, timing_ms, error))
    
    def retrieve(self) -> None:
        start_time = time.time()
        terms = self.__context.get_terms()
        inputs = f"Size of terms {len(terms)} Terms: {terms}"
        outputs_summary = ""
        error = None
        
        try:
            from src.retrieval.keyword_retriever import KeywordRetriever
            
            if self.__config.get_retriever_type() == "KeywordRetriever":
                self._retriever = KeywordRetriever(self.__config.get_top_k())
            else:
                raise IllegalArgumentError(f"Unknown retriever type: {self.__config.get_retriever_type()}")
            
            hits = self._retriever.retrieve(terms, self.__context.get_chunk_store())
            self.__context.set_retrieved_hits(hits)
            outputs_summary = f"Number of hits: {len(hits)} retrievedHits: {hits}"
        except Exception as e:
            error = str(e)
            raise
        finally:
            timing_ms = (time.time() - start_time) * 1000
            self.__trace_bus.publish(TraceEvent("retrieve", inputs, outputs_summary, timing_ms, error))
    
    def rerank(self) -> None:
        
        start_time = time.time()
        terms = self.__context.get_terms()
        hits = self.__context.get_retrieved_hits()
        inputs = f"Size of retrievedHits: {len(hits)} Hits: {hits}"
        outputs_summary = ""
        error = None
        
        try:
            from src.reranker.simple_reranker import SimpleReranker
            from src.data.chunk_loader import ChunkLoader
            
            if self.__config.get_reranker_type() == "SimpleReranker":
                proximity_window = 15
                proximity_bonus = 5
                title_boost = 3
                self._reranker = SimpleReranker(proximity_window, proximity_bonus, title_boost)
            else:
                raise IllegalArgumentError(f"Unknown reranker type: {self.__config.get_reranker_type()}")
            
            chunk_loader = ChunkLoader()
            chunk_store = chunk_loader.load_chunks(self.__config.get_chunk_path())
            
            reranked_hits = self._reranker.rerank(terms, hits, chunk_store)
            self.__context.set_reranked_hits(reranked_hits)
            outputs_summary = f"Size of rerankedHits: {len(reranked_hits)} hits: {reranked_hits}"
        except Exception as e:
            error = str(e)
            raise
        finally:
            timing_ms = (time.time() - start_time) * 1000
            self.__trace_bus.publish(TraceEvent("rerank", inputs, outputs_summary, timing_ms, error))
    
    def answer(self) -> None:
        start_time = time.time()
        reranked_hits = self.__context.get_reranked_hits()
        inputs = f"Number of hits: {len(reranked_hits)} rerankedHits: {reranked_hits}"
        outputs_summary = ""
        error = None
        
        try:
            from src.answer.template_answer_agent import TemplateAnswerAgent
            
            if self.__config.get_answer_agent_type() == "TemplateAnswerAgent":
                self._answer_agent = TemplateAnswerAgent()
            else:
                raise IllegalArgumentError(f"Unknown answer agent type: {self.__config.get_answer_agent_type()}")
            
            query_terms = self.__context.get_terms()
            generated_answer = self._answer_agent.answer(query_terms, reranked_hits, self.__context.get_chunk_store())
            self.__context.set_final_answer(generated_answer)
            outputs_summary = f"Answer: {generated_answer}"
        except Exception as e:
            error = str(e)
            raise
        finally:
            timing_ms = (time.time() - start_time) * 1000
            self.__trace_bus.publish(TraceEvent("answer", inputs, outputs_summary, timing_ms, error))
    
    def load_stopwords(self, stopwords_path: Path) -> Set[str]:
        try:
            stopwords = set()
            in_stopwords = False
            
            with open(stopwords_path, 'r', encoding='utf-8') as f:
                for line in f:
                    trimmed = line.strip()
                    
                    if not trimmed or trimmed.startswith("#"):
                        continue
                    
                    if trimmed == "stop_words:":
                        in_stopwords = True
                        continue
                    
                    if in_stopwords and trimmed.startswith("-"):
                        word = trimmed[1:].strip()
                        if word.startswith('"') and word.endswith('"'):
                            word = word[1:-1]
                        stopwords.add(word)
            
            return stopwords
        except Exception as e:
            raise RuntimeError(f"Failed to load stopwords from: {stopwords_path}") from e


class IllegalArgumentError(Exception):
    pass
